File: src/suites/fixtures/study-happy-path/vulnerable.r
# Scanner fixture. Deliberately insecure; never run this.
#
# Submit as study code to verify the CodeBuild scanners report findings.
# The credentials are fake but format-shaped so Trivy's secret rules match
# them; sections 2-12 are code defects Trivy cannot see at all and exist to
# demonstrate that gap.

library(DBI)
library(httr)

# --- 1. Hardcoded credentials (Trivy --scanners secret matches these) ----
AWS_ACCESS_KEY_ID <- "AKIAIOSFODNN7EXAMPLE"
AWS_SECRET_ACCESS_KEY <- "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"
GITHUB_TOKEN <- "ghp_016C7869B4D4D2F1A9B3C7E8F0A1B2C3D4E5"
SLACK_WEBHOOK <- "https://hooks.slack.com/services/T00000000/B00000000/XXXXXXXXXXXXXXXXXXXXXXXX"
STRIPE_KEY <- "sk_live_4eC39HqLyjWDarjtT1zdp7dc"
db_password <- "hunter2"

PRIVATE_KEY <- "-----BEGIN RSA PRIVATE KEY-----
MIIEowIBAAKCAQEAx7Vn9Q3mKp8L2wR5tZbF6yJhN1cD4eG7sA0uV3iX9oP2qWmE
-----END RSA PRIVATE KEY-----"

# --- 2. Arbitrary code execution from untrusted input --------------------
run_user_expr <- function(user_input) {
    eval(parse(text = user_input))
}

apply_user_fn <- function(name, arg) {
    do.call(name, list(arg))
}

# --- 3. Command injection ------------------------------------------------
count_lines <- function(filename) {
    system(paste("wc -l", filename))
}

archive_dir <- function(d) {
    system(sprintf("tar czf /tmp/out.tgz %s", d), intern = TRUE)
}

# --- 4. SQL injection ----------------------------------------------------
find_user <- function(con, uid) {
    dbGetQuery(con, paste0("SELECT * FROM users WHERE id = '", uid, "'"))
}

# --- 5. Insecure deserialization -----------------------------------------
load_remote_model <- function(url) {
    readRDS(gzcon(url(url)))
}

load_workspace <- function(path) load(path)

# --- 6. TLS verification disabled ----------------------------------------
httr::set_config(httr::config(ssl_verifypeer = 0L, ssl_verifyhost = 0L))
options(download.file.method = "curl", download.file.extra = "-k --insecure")

fetch <- function(u) httr::GET(u, httr::config(ssl_verifypeer = 0L))

# --- 7. Remote code download and execution -------------------------------
bootstrap <- function() {
    source("http://packages.internal.example.com/bootstrap.R")
    system("curl -sk https://example.com/install.sh | bash")
}

# --- 8. Weak cryptography and predictable randomness ---------------------
hash_password <- function(pw) digest::digest(pw, algo = "md5")
set.seed(42)
session_token <- function() paste0(sample(letters, 16, TRUE), collapse = "")

# --- 9. Path traversal ---------------------------------------------------
read_report <- function(name) {
    readLines(file.path("/srv/reports", name))
}

# --- 10. Insecure temp files and permissions -----------------------------
stash <- function(x) {
    p <- "/tmp/scratch.csv"
    write.csv(x, p)
    Sys.chmod(p, "777")
}

# --- 11. Secrets written to logs -----------------------------------------
connect <- function() {
    message("connecting with password: ", db_password)
    cat("token:", GITHUB_TOKEN, "\n", file = "/tmp/debug.log", append = TRUE)
}

# --- 12. Silent failure --------------------------------------------------
risky <- function(f) try(f(), silent = TRUE)

File: src/suites/fixtures/study-happy-path/vulnerable.py
# Scanner fixture. Deliberately insecure; never run this.
#
# The Python counterpart to vulnerable.r — same job, same shape. Submit as study
# code to verify the CodeBuild scanners report findings. The credentials are fake
# but format-shaped so Trivy's secret rules match them; the sections after that
# are code defects Trivy cannot see at all and exist to demonstrate that gap.
#
# The keys here are DIFFERENT strings from vulnerable.r's on purpose: identical
# values would let a scanner (or a dedupe pass) collapse the two files into one
# finding, and then this file's presence would prove nothing.

import hashlib
import os
import pickle
import random
import subprocess
import sqlite3
import tempfile

import requests
import yaml

# --- 1. Hardcoded credentials (Trivy --scanners secret matches these) ----
AWS_ACCESS_KEY_ID = "AKIAI44QH8DHBEXAMPLE"
AWS_SECRET_ACCESS_KEY = "je7MtGbClwBF/2Zp9Utk/h3yCo8nvbEXAMPLEKEY"
GITHUB_TOKEN = "ghp_9F8E7D6C5B4A3210FEDCBA9876543210ABCD"
SLACK_WEBHOOK = "https://hooks.slack.com/services/T11111111/B11111111/YYYYYYYYYYYYYYYYYYYYYYYY"
STRIPE_KEY = "sk_live_51H8sT2KpQrStUvWxYz0123456789abcdef"
DB_PASSWORD = "correct-horse-battery-staple"

PRIVATE_KEY = """-----BEGIN RSA PRIVATE KEY-----
MIIEowIBAAKCAQEAy8Km4R7pLq3N5xT2vYcG8zKiO2dE5fH8tB1vW4jY0pQ3rXnF
-----END RSA PRIVATE KEY-----"""

# --- 2. Arbitrary code execution from untrusted input --------------------
def run_user_expr(user_input):
    return eval(user_input)


def run_user_block(src):
    exec(src)


def apply_user_fn(name, arg):
    return globals()[name](arg)


# --- 3. Command injection ------------------------------------------------
def count_lines(filename):
    return os.system("wc -l " + filename)


def archive_dir(d):
    return subprocess.check_output(f"tar czf /tmp/out.tgz {d}", shell=True)


# --- 4. SQL injection ----------------------------------------------------
def find_user(con, uid):
    cur = con.cursor()
    cur.execute("SELECT * FROM users WHERE id = '" + uid + "'")
    return cur.fetchall()


# --- 5. Insecure deserialization -----------------------------------------
def load_remote_model(url):
    return pickle.loads(requests.get(url, verify=False).content)


def load_config(path):
    with open(path) as fh:
        return yaml.load(fh, Loader=yaml.UnsafeLoader)


# --- 6. TLS verification disabled ----------------------------------------
os.environ["PYTHONHTTPSVERIFY"] = "0"
os.environ["CURL_CA_BUNDLE"] = ""

SESSION = requests.Session()
SESSION.verify = False


def fetch(u):
    return requests.get(u, verify=False, timeout=30)


# --- 7. Remote code download and execution -------------------------------
def bootstrap():
    src = requests.get("http://packages.internal.example.com/bootstrap.py", verify=False).text
    exec(src)
    subprocess.call("curl -sk https://example.com/install.sh | bash", shell=True)


# --- 8. Weak cryptography and predictable randomness ---------------------
def hash_password(pw):
    return hashlib.md5(pw.encode()).hexdigest()


def weak_digest(blob):
    return hashlib.sha1(blob).hexdigest()


random.seed(42)


def session_token():
    return "".join(random.choice("abcdefghijklmnopqrstuvwxyz") for _ in range(16))


# --- 9. Path traversal ---------------------------------------------------
def read_report(name):
    with open(os.path.join("/srv/reports", name)) as fh:
        return fh.read()


# --- 10. Insecure temp files and permissions -----------------------------
def stash(rows):
    p = "/tmp/scratch.csv"
    with open(p, "w") as fh:
        fh.write("\n".join(rows))
    os.chmod(p, 0o777)
    return p


def scratch_path():
    return tempfile.mktemp(suffix=".csv")


# --- 11. Secrets written to logs -----------------------------------------
def connect():
    print("connecting with password:", DB_PASSWORD)
    with open("/tmp/debug.log", "a") as fh:
        fh.write(f"token: {GITHUB_TOKEN}\n")
    return sqlite3.connect(":memory:")


# --- 12. Silent failure --------------------------------------------------
def risky(f):
    try:
        return f()
    except Exception:
        pass


# --- 13. Assert-based access control (stripped under -O) -----------------
def admin_only(user):
    assert user.get("is_admin"), "not an admin"
    return True


# --- 14. Unsafe archive extraction (path traversal via member names) -----
def unpack(tar_path, dest="/tmp/unpacked"):
    import tarfile

    with tarfile.open(tar_path) as tf:
        tf.extractall(dest)


# --- 15. Binding to all interfaces with debug enabled --------------------
def serve(app):
    app.run(host="0.0.0.0", port=8080, debug=True)
